### ISSUE TITLE
Duplicate task execution

### ISSUE THREAD

**Issue by bhugot** [microsoft/durabletask-mssql#132]
Hello, with a standard configuration of the provider, we juste found a use case where same activity was run twice without any trace in the history. Only trace we found was the double call in our appinsight. 
I will try to reproduce it. (the server was not at all under heavy stress)

**Comment by cgillum** [comment]
Were the duplicate calls on the same VM instance or different VM instances? Also, how much time was there between the two invocations. In most cases, this would indicate an automatic retry caused by a process failure.

**Comment by bhugot** [comment]
I think it's matching lock has the duration took about 2 min more than usual

**Comment by bhugot** [comment]
it's different VM instances

**Comment by gsimardimt** [comment]
Is this issue still active ? We also found a case of activities being run twice. Our function app is in consumption mode and the interval between the two invocations is about 5 minutes.

**Comment by cgillum** [comment]
Yes, still active. However, it should be noted that some duplicate task executions can be expected, particularly when VMs come in and out of rotation. This is especially true when running function apps in the consumption plan, where app instances can be created and removed within seconds. There is _potentially_ some work that can be done to make the transitions more graceful and reduce the likelihood of duplicate executions. More investigation would be required to see what's possible.

**Comment by gsimardimt** [comment]
It has been a month without any feedback. In my view, this is a significant issue. If we cannot trust that an activity will execute as expected, what is the point of using Durable Tasks?

**Comment by cgillum** [comment]
> If we cannot trust that an activity will execute as expected, what is the point of using Durable Tasks?

I wonder if there's a misunderstanding - Durable Tasks guarantees at-least-once execution. I don't believe there was any report about activities not executing.

**Comment by gsimardimt** [comment]
Absolutely, I didn’t realize that every activity must be idempotent. So, is there nothing in place to prevent that behavior?

**Comment by cgillum** [comment]
There's nothing in place to prevent duplicate executions. One thing we're considering, however, is introducing metadata to the activity context that would tell you if it might be a duplicate invocation. You'd still need to ensure that your logic is idempotent, but having hints might allow you to do idempotency checks selectively and reduce overhead.
